pkg/kubectl/plugins/loader.go

```
@@ -0,0 +1,193 @@
+		return nil, err
+	}
+	if !stat.IsDir() {
+		return nil, fmt.Errorf("not a directory: %s", l.Directory)
+	}
+
+	base, err := filepath.Abs(l.Directory)
+	if err != nil {
+		return nil, err
+	}
+
+	// read the base directory tree searching for plugin descriptors
+	// fails silently (descriptors unable to be read or unmarshalled are logged but skipped)
+	err = filepath.Walk(base, func(path string, fileInfo os.FileInfo, walkErr error) error {
+		if fileInfo.IsDir() || fileInfo.Name() != PluginDescriptorFilename || walkErr != nil {
+			return nil
+		}
+
+		file, err := ioutil.ReadFile(path)
+		if err != nil {
+			glog.V(6).Infof("Unable to read plugin descriptor %s: %v", path, err)
+			return nil
+		}
+
+		plugin := &Plugin{}
+		if err := yaml.Unmarshal(file, plugin); err != nil {
+			glog.V(6).Infof("Unable to unmarshal plugin descriptor %s: %v", path, err)
+			return nil
+		}
+
+		if err := plugin.Validate(); err != nil {
+			glog.V(6).Infof("%v", err)
+			return nil
+
+		}
+
+		plugin.Dir = filepath.Dir(path)
+		plugin.DescriptorName = fileInfo.Name()
+
+		glog.V(6).Infof("Plugin loaded: %s", plugin.Name)
+		list = append(list, plugin)
+
+		return nil
+	})
+
+	return list, err
+}
+
+// UserDirPluginLoader is a PluginLoader that loads plugins from the
+// "plugins" directory under the user's kubeconfig dir (usually "~/.kube/plugins/").
+func UserDirPluginLoader() PluginLoader {
+	dir := filepath.Join(clientcmd.RecommendedConfigDir, "plugins")
+	return &DirectoryPluginLoader{
+		Directory: dir,
+	}
+}
+
+// PathFromEnvVarPluginLoader is a PluginLoader that loads plugins from one or more
+// directories specified by the provided env var name. In case the env var is not
+// set, the PluginLoader just loads nothing. A list of subdirectories can be provided,
+// which will be appended to each path specified by the env var.
+func PathFromEnvVarPluginLoader(envVarName string, subdirs ...string) PluginLoader {
+	env := os.Getenv(envVarName)
+	if len(env) > 0 {
```

Revers this condition:
```
if len(env) == 0 {
    return &DummyPluginLoader{}
}
loader := MultiPluginLoader{}
...
```
The rule is fail fast, and try to avoid doing unnecessary indention ;)